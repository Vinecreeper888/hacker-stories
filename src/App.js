import React from 'react';
import axios from 'axios';
// import cs from 'classnames';
import styled from 'styled-components';
import List from './List';
import SearchForm from './SearchForm';

// interface StoriesFetchInitAction {
//   type: 'STORIES_FETCH_INIT';
// }

// interface StoriesFetchSuccessAction {
//   type: 'STORIES_FETCH_SUCCESS';
//   payload: Stories;
// }

// interface StoriesFetchFailureAction {
//   type: 'STORIES_FETCH_FAILURE';
// }

// interface StoriesRemoveAction {
//   type: 'REMOVE_STORY';
//   payload: Story;
// }

// type StoriesState = {
//   data: Stories;
//   isLoading: boolean;
//   isError: boolean;
// };

//const API_ENDPOINT = 'https://hn.algolia.com/api/v1/search?query=';
const API_BASE = 'https://hn.algolia.com/api/v1';
const API_SEARCH = '/search';
const PARAM_SEARCH = 'query=';
const PARAM_PAGE = 'page=';

//styled components
const StyledContainer = styled.div`
  height: 100vw;
  padding: 20px;
  
  background: #83a4d4;
  background: linear-gradient(to left, #b6fbff, #83a4d4);
  color: #171212;
`;

const StyledHeadlinePrimary = styled.h1`
  font-size: 48px;
  font-weight: 300;
  letter-spacing: 2px;
`;

// const StyledItem = styled.div`
//   display: flex;
//   align-items: center;
//   padding-bottom: 5px;
// `;

// const StyledColumn = styled.span`
//   padding: 0 5px;
//   white-space: nowrap;
//   overflow: hidden;
//   white-space: nowrap;
//   text-overflow: ellipse;

//   a {
//     color: inherit;
//   }

//   width: ${props => props.width};
// `;

// const StyledButtonSmall = styled(StyledButton)`
//   padding: 5px;
// `;

// type StoriesAction = 
//   | StoriesFetchInitAction
//   | StoriesFetchSuccessAction
//   | StoriesFetchFailureAction
//   | StoriesRemoveAction;

//advanced state mgmt
const storiesReducer = (
  state, 
  action
  ) => {
  switch (action.type) {
    case 'STORIES_FETCH_INIT':
      return {
        ...state,
        isLoading: true,
        isError: false
      };
    case 'STORIES_FETCH_SUCCESS':
      return {
        ...state,
        isLoading: false,
        isError: false,
        data: 
          action.payload.page === 0
            ? action.payload.list
            : state.data.concat(action.payload.list),
        page: action.payload.page,
      };
    case 'STORIES_FETCH_FAILURE':
      return {
        ...state,
        isLoading: false,
        isError: true,
      };
    case 'REMOVE_STORY':
      return {
        ...state,
        data: state.data.filter(
          story => action.payload.objectID !== story.objectID
        ),
      };
    default:
      throw new Error();  
  }
};
  
//custom hook
const useSemiPersistentState = (
  key, 
  initialState
)  => {
  //const isMounted = React.useRef(false);

  const [value, setValue] = React.useState(
    localStorage.getItem(key) || initialState
  );

  React.useEffect(() => {
    // if(!isMounted.current) {
    //   isMounted.current = true;
    // } else {
    //   console.log('A');
      localStorage.setItem(key,value);
    //} 
  },[value,key]);
  
  return [value, setValue];
};


// const getSumComments = stories => {

//   return stories.data.reduce(
//     (result,value) => result + value.num_comments,0
//   );
// };

const extractSearchTerm = url => 
  url
    .substring(url.lastIndexOf('?')+1, url.lastIndexOf('&'))
    .replace(PARAM_SEARCH,'');

const getLastSearches = urls => 
  urls
  .reduce((result,url,index) => {
    const searchTerm = extractSearchTerm(url);

    if(index === 0) {
      return result.concat(searchTerm);
    }

    const previousSearchTerm = result[result.length - 1];

    if(searchTerm === previousSearchTerm) {
      return result;
    } else {
      return result.concat(searchTerm);
    }
  },[])
  .slice(-6)
  .slice(0,-1)
  .map(url => extractSearchTerm(url));

const getUrl = (searchTerm,page) => 
  `${API_BASE}${API_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}`; 



const App = () => {
  
  const [searchTerm, setSearchTerm] = useSemiPersistentState('search','');

  const [urls, setUrls] = React.useState([
    getUrl(searchTerm,0)
  ]);

  const [stories, dispatchStories] = React.useReducer(
    storiesReducer,
    {data: [], page: 0, isLoading: false, isError: false}
  );

  const handleSearchInput = (
    event
  ) => {
    setSearchTerm(event.target.value);
  };

  const handleSearchSubmit = (
    event
    ) => {
      handleSearch(searchTerm,0);
      
      event.preventDefault();
    };
    
    const handleLastSearch = searchTerm => {
      setSearchTerm(searchTerm,0);
    };

    const handleSearch = (searchTerm,page) => {
      const url = getUrl(searchTerm,page);
      setUrls(urls.concat(url));
    };

    const handleFetchStories = React.useCallback(async () => {
  
    dispatchStories({type: 'STORIES_FETCH_INIT'});

    //for error handling, use try catch
    try {
      const lastUrl = urls[urls.length - 1];
      const result = await axios.get(lastUrl);
      dispatchStories({
        type: 'STORIES_FETCH_SUCCESS',
        payload: {
        list: result.data.hits,
        page: result.data.page,
        }
      });
    } catch {
      dispatchStories({type: 'STORIES_FETCH_FAILURE'});
    }
  },[urls]);

  React.useEffect(() => {
    handleFetchStories();
  },[handleFetchStories])

  const handleRemoveStory = (item) => {
    dispatchStories({
      type: 'REMOVE_STORY',
      payload: item,
    });
  };

  React.useEffect(() => {
    localStorage.setItem('search',searchTerm);
  },[searchTerm]);

  const handleMore = () => {
    const lastUrl = urls[urls.length - 1];
    const searchTerm = extractSearchTerm(lastUrl);
    handleSearch(searchTerm,stories.page + 1);
  };

  const lastSearches = getLastSearches(urls);

  return (
    <StyledContainer>
      <StyledHeadlinePrimary>My Hacker Stories</StyledHeadlinePrimary>
      
      <SearchForm
        searchTerm={searchTerm}
        onSearchInput={handleSearchInput}
        onSearchSubmit={handleSearchSubmit}
      /><br/>

      <LastSearches 
        lastSearches={lastSearches}
        onLastSearch={handleLastSearch}
      />

      {lastSearches.map((searchTerm, index) => (
        <button
          key={searchTerm+index}
          type="button"
          onClick={() => handleLastSearch(searchTerm)}
        >
          {searchTerm}
        </button>
      ))}

      {/* <hr/> */}

      <List list={stories.data} onRemoveItem={handleRemoveStory}/>
  
      {stories.isError && <p>Something went wrong ...</p>}
  
      {stories.isLoading ? (
        <p>Loading...</p>
      ): (
        <button type="button" onClick={handleMore}>
          More
        </button>
      )}
    </StyledContainer>
  );
};

const LastSearches = ({lastSearches, onLastSearch}) => (
  <>
    {lastSearches.map((searchTerm,index) => (
      <button
        key={searchTerm}
        type="button"
        onClick={() => onLastSearch(searchTerm)}
      >
        {searchTerm}
      </button>
    ))}
  </>
);

// type SearchFormProps = {
//   searchTerm: string;
//   onSearchInput: (event: React.ChangeEvent<HTMLInputElement>) => void;
//   onSearchSubmit: (event: React.FormEvent<HTMLFormElement>) => void;
// };

//console.log('B:App');

//const sumComments = React.useMemo(() => getSumComments(stories),[stories]);

// type InputWithLabelProps = {
//   id: string;
//   value: string;
//   type?: string;
//   onInputChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
//   isFocused?: boolean;
//   children: React.ReactNode;
// };

export default App;
