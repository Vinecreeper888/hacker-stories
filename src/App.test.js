
// ------------- SOME INFO ON JEST SYNTAX -------------
// a) Jest consists of test suites(describe)
// b) test cases (it)
// c) assertions (expect) that turn out green or red

import { render, screen } from '@testing-library/react';
import App, {Item, List, SearchForm, InputWithLabel} from './App';
import renderer from 'react-test-renderer';
import axios from 'axios';

jest.mock('axios');

describe('Item', () => {
  const item = {
    title: 'React',
    url: 'https://reactjs.org',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
  };
  const handleRemoveItem = jest.fn();

  let component;

  beforeEach(() => {
    component = renderer.create(
      <Item item={item} onRemoveItem={handleRemoveItem}/>
    );
  });

  it('renders all properties',() => {
    const component = renderer.create(<Item item={item}/>);
  });
  it('calls onRemoveItem on button click',() => {
    const handleRemoveItem = jest.fn();

    const component = renderer.create(
      <Item item={item} onRemoveItem={handleRemoveItem}/>
    );
    component.root.findByType('button').props.onClick();

    expect(handleRemoveItem).toHaveBeenCalledTimes(1);
    expect(handleRemoveItem).toHaveBeenCalledWith(item);

    expect(component.root.findAllByType(Item).length).toEqual(1);
  });

  //snapshot testing
  test('renders snapshot',() => {
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});


describe('List', () => {
  const list = [
    {
      title: 'React',
      url: 'https://reactjs.org/',
      author: 'Jordan Walke',
      num_comments: 3,
      points: 4,
      objectID: 0,
    },
    {
      title: 'Redux',
      url: 'https://redux.js.org/',
      author: 'Dan Abramov, Andrew Clark',
      num_comments: 2,
      points: 5,
      objectID: 1,
    }
  ];

  it('renders two items',() => {
    const component = renderer.create(<List list={list}/>);

    expect(component.root.findAllByType(Item).length).toEqual(2);
  });
});

describe('SearchForm',() => {
  const searchFormProps = {
    searchTerm: 'React',
    onSearchInput: jest.fn(),
    onSearchSubmit: jest.fn(),
  };
  let component;

  beforeEach(() => {
    component = renderer.create(<SearchForm{...searchFormProps}/>);
  });

  it('renders the input field with its value',() => {
    const value = component.root.findByType('input').props.value;

    expect(value).toEqual('React');
  });

  it('changes the input field',() => {
    const pseudoEvent = {target: 'Redux'};

    component.root.findByType('input').props.onChange(pseudoEvent);

    expect(searchFormProps.onSearchInput).toHaveBeenCalledTimes(1);
    expect(searchFormProps.onSearchInput).toHaveBeenLastCalledWith(pseudoEvent);
  });

  it('submits the form',() => {
    const pseudoEvent = {};

    component.root.findByType('form').props.onSubmit(pseudoEvent);

    expect(searchFormProps.onSearchSubmit).toHaveBeenCalledTimes(1);
    expect(searchFormProps.onSearchSubmit).toHaveBeenLastCalledWith(pseudoEvent);
  });

  it('disables the button and prevents submit',() => {
    component.update(
      <SearchForm {...searchFormProps} searchTerm=""/>
    );

    expect(
      component.root.findByType('button').props.disabled
    ).toBeTruthy();
  })
});

describe('App',() => {
  it('succeeds fetching data with a list', async () => {
    const list = [
      {
        title: 'React',
        url: 'https://reactjs.org/',
        author: 'Jordan Walke',
        num_comments: 3,
        points: 4,
        objectID: 0,
      },
      {
        title: 'Redux',
        url: 'https://redux.js.org/',
        author: 'Dan Abramov, Andrew Clark',
        num_comments: 2,
        points: 5,
        objectID: 1,
      },
    ];

    const promise = Promise.resolve({
      data: {
        hits: list,
      },
    });

    axios.get.mockImplementationOnce(() => promise);

    let component;
    
    await renderer.act(async() => {
      component = renderer.create(<App/>);
    }); 

    expect(component.root.findByType(List).props.list).toEqual(list);
  });
});
  