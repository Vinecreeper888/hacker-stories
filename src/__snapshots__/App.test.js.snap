// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item renders snapshot 1`] = `
<div
  className="sc-dlfnbm gebkuo"
>
  <span
    className="sc-hKgILt jDLFvj"
    width="40%"
  >
    <a
      href="https://reactjs.org"
    >
      React
    </a>
  </span>
  <span
    className="sc-hKgILt jDKSoc"
    width="30%"
  >
    Jordan Walke
  </span>
  <span
    className="sc-hKgILt jDLroW"
    width="10%"
  >
    3
  </span>
  <span
    className="sc-hKgILt jDLroW"
    width="10%"
  >
    4
  </span>
  <span
    className="sc-hKgILt jDLroW"
    width="10%"
  >
    <button
      className="sc-eCssSg sc-jSgupP fdMRLi fqYsgG"
      onClick={[Function]}
      type="button"
    >
      Dismiss
      <svg
        height="18px"
        width="18px"
      >
        check.svg
      </svg>
    </button>
  </span>
</div>
`;
